src/dlfcn.c

```diff
@@ -567,20 +567,26 @@ static void fill_module_info( void *addr, Dl_info *info )
     HMODULE hModule;
     DWORD dwSize;
 
+    info->dli_fname = NULL;
     if( !GetModuleHandleExA( GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, addr, &hModule ) || !hModule )
     {
-        info->dli_fname = NULL;
         info->dli_fbase = NULL;
         return;
     }
 
     info->dli_fbase = (void *) hModule;
 
     dwSize = GetModuleFileNameA( hModule, module_filename, sizeof( module_filename ) );
-    if( dwSize > 0 && dwSize < sizeof( module_filename ) )
-        info->dli_fname = module_filename;
-    else
-        info->dli_fname = NULL;
+    if( dwSize == 0 )
+        return;
+    /* Windows XP */
+    if( dwSize == sizeof( module_filename ) && GetLastError() == ERROR_SUCCESS )
+        return;
+    /* > Windows XP */
+    if( dwSize == sizeof( module_filename ) && GetLastError() == ERROR_INSUFFICIENT_BUFFER )
+        return;
+
+    info->dli_fname = module_filename;

```

Comment:
There is no need to complicate it.

> If the function succeeds, the return value is the length of the string that is copied to the buffer, in characters, not including the terminating null character.

So if returned size is not smaller then size of buffer then overflow happened.

I think you have it correct without this change.